shrinklvm.py

```diff
@@ -0,0 +1,1012 @@
+    bytes, moves the end backward by end_space bytes, and then resizes the
+    partition to fit. The start is moved forward precisely to ensure that
+    alignment is respected. This is done by increasing the size of the metadata
+    area. The end_space is rounded up to the nearest extent.
+    '''
+
+    logger.log('Reading LVM header...')
+    lvm = read_header(partitionDevice(drive, partno))
+    logger.log('Original LVM config: {}'.format(lvm.text_config))
+
+    logger.log('Parsing LVM config...')
+    parse_config(lvm)
+
+    logger.log('Checking LVM device...')
+    check_lvm(lvm)
+
+    # Create a sorted allocation table of the segments
+    alloc_table = lvm_to_alloc_table(lvm)
+    logger.log('Initial allocation table: {}'.format(alloc_table))
+
+    extent_size = int(lvm.vgs[0].get_config('extent_size')) * SECTOR_SIZE
+    pe_count = int(lvm.vgs[0].pvs[0].get_config('pe_count'))
+    start_shrunk_extents = int(math.ceil(start_space / float(extent_size)))
+    end_shrunk_extents = int(math.ceil(end_space / float(extent_size)))
+    regions = [ReservedRegion(0, start_shrunk_extents)]
+    if end_shrunk_extents > 0:
+        regions.append(ReservedRegion(pe_count - end_shrunk_extents, end_shrunk_extents))
+
+    # If we're shrinking the start by an amount that is not a multiple of the extent size
+    # increase the size of the metadata area to compensate
+    extra_metadata_sectors = (start_shrunk_extents * extent_size - start_space) / SECTOR_SIZE
+    logger.log('Increasing metadata size by {} sectors'.format(extra_metadata_sectors))
+
+    logger.log('Reserved regions: {}'.format(regions))
+
+    move_ops = move_segment_allocations(alloc_table, regions, pe_count - 1)
+
+    logger.log('Planned move operations: {}'.format(move_ops))
+    logger.log('Planned final allocation table: {}'.format(alloc_table))
+
+    progress_callback(5)
+
+    # Up to this point, the disk has not been modified.
+    # Now that checks are complete and we have a plan,
+    # begin the destructive part.
+
+    # Emit progress accurately. Pre-move and post-move operations take 5% each
+    # with moving data taking 90% of the progress.
+    bytes_to_move = sum([i[2] * extent_size for i in move_ops])
+    def get_move_ops_progress_cb():
+        moved_bytes = [0]
+        progress = [-1]
+
+        def subprogress_callback(n):
+            moved_bytes[0] += n
+            new_progress = 5 + int(float(moved_bytes[0]) / bytes_to_move * 90)
+            if new_progress != progress[0]:
+                progress[0] = new_progress
+                progress_callback(new_progress)
+
+        return subprogress_callback
+
+    move_ops_cb = get_move_ops_progress_cb()
+    for op in move_ops:
+        move_extents(lvm, move_ops_cb, *op)
+
+    # A crash up to this point shouldn't affect anything because we've only
+    # written into free unallocated extents. Now enter the critical region and
+    # commit the changes.
```

Comment:
Potentially we could update the layout in the current partition so the next write will also update free space. But not a big issue I suppose

Reply:
Sorry, I'm not sure what you mean. Could you clarify please?

Reply:
More of an improvement to avoid data loss... or better make possible a recovery. Now you moved the extents to free space. If you update the LVM headers at the current position and a restart/crash happens (for whatever reason) the LVs will still have the same data. If you don't have a restart (which should be always) you will be writing the new LVM headers in the new position which are now marked as free. Now, if you have a restart/crash here the LVM will point to old headers so LVs will still have the same data. Yes, at the point of rebuilding the partition schema then you could loose data, but rewriting the LVM headers will decrease the critical region.

As said, mainly paranoia mode here...

Reply:
OK, that makes sense. I'll add it.

To reduce the window further, I'll also update the code to do the shrink partition stuff earlier and just let it do the final `commit()` at the end.